File: android-host/app/src/main/java/com/huddle/host/runtime/plugins/UndercoverPlugin.kt
package com.huddle.host.runtime.plugins

import android.content.Context
import com.huddle.host.data.UndercoverWordStore
import com.huddle.host.runtime.GamePlugin
import com.huddle.host.runtime.Player
import com.huddle.host.runtime.PluginResult
import org.json.JSONArray
import org.json.JSONObject

/**
 * Who-is-the-undercover MVP: deal private words, host reveals.
 * Word pairs persist in [UndercoverWordStore]; used pairs are marked after deal.
 */
class UndercoverPlugin(
    private val appContext: Context
) : GamePlugin {
    override val id: String = "undercover"

    private var hostId: String? = null
    private var undercoverCount: Int = 1
    private var pairId: String? = null
    private var civilianWord: String = ""
    private var undercoverWord: String = ""
    private var phase: String = "secret" // secret | revealed
    private var finished: Boolean = false
    private var started: Boolean = false

    /** playerId -> role ("civilian" | "undercover") */
    private val roles = linkedMapOf<String, String>()
    private val names = linkedMapOf<String, String>()
    private val seats = linkedMapOf<String, Int>()

    override fun onStart(players: List<Player>, hostId: String): PluginResult {
        reset()
        val minNeed = UndercoverWordStore.minPlayers(appContext)
        if (players.size < minNeed) {
            return PluginResult.Err("谁是卧底至少需要 ${minNeed} 人入座")
        }
        val count = UndercoverWordStore.undercoverCount(appContext)
        if (count < 1 || count >= players.size) {
            return PluginResult.Err("卧底人数须在 1 到 ${players.size - 1} 之间，请先在配置里调整")
        }
        val pair = UndercoverWordStore.pickUnused(appContext)
            ?: return PluginResult.Err("词库没有可用词组，请先在配置里添加或重新加入")

        this.hostId = hostId
        undercoverCount = count
        pairId = pair.id
        civilianWord = pair.civilian
        undercoverWord = pair.undercover

        val shuffled = players.shuffled()
        val undercoverIds = shuffled.take(count).map { it.id }.toSet()
        for (p in players) {
            roles[p.id] = if (p.id in undercoverIds) "undercover" else "civilian"
            names[p.id] = p.name
            seats[p.id] = p.seat ?: 0
        }

        UndercoverWordStore.markUsed(appContext, pair.id)
        phase = "secret"
        finished = false
        started = true

        return PluginResult.Ok(
            system = "词语已发放 · ${players.size} 人 · 卧底 $count 人 · 线下讨论后由房主揭晓"
        )
    }

    override fun onAction(playerId: String, payload: JSONObject): PluginResult {
        if (!started) return PluginResult.Err("对局未开始")
        if (finished || phase == "revealed") return PluginResult.Err("已经揭晓")
        val op = payload.optString("op", "")
        if (op != "reveal") {
            return PluginResult.Err("未知操作")
        }
        if (playerId != hostId) {
            return PluginResult.Err("只有房主可以揭晓")
        }
        phase = "revealed"
        finished = true
        return PluginResult.Ok(
            finished = true,
            system = "身份已揭晓"
        )
    }

    override fun snapshot(): JSONObject = buildSnapshot(viewerId = null)

    override fun snapshotFor(playerId: String): JSONObject = buildSnapshot(viewerId = playerId)

    private fun buildSnapshot(viewerId: String?): JSONObject {
        val revealed = phase == "revealed"
        val arr = JSONArray()
        for ((pid, role) in roles) {
            val o = JSONObject()
                .put("id", pid)
                .put("name", names[pid] ?: "玩家")
                .put("seat", seats[pid] ?: 0)
            if (revealed) {
                o.put("role", role)
                o.put("word", if (role == "undercover") undercoverWord else civilianWord)
            }
            arr.put(o)
        }

        val snap = JSONObject()
            .put("phase", phase)
            .put("undercoverCount", undercoverCount)
            .put("players", arr)
            .put("finished", finished)

        if (viewerId != null && roles.containsKey(viewerId)) {
            val myRole = roles[viewerId]!!
            val myWord = if (myRole == "undercover") undercoverWord else civilianWord
            snap.put("myWord", myWord)
            // Identity only after reveal; during secret only the word is shown.
            if (revealed) {
                snap.put("myRole", myRole)
            }
        }

        if (revealed) {
            snap.put("civilianWord", civilianWord)
            snap.put("undercoverWord", undercoverWord)
        }

        return snap
    }

    override fun isFinished(): Boolean = finished

    override fun reset() {
        hostId = null
        undercoverCount = 1
        pairId = null
        civilianWord = ""
        undercoverWord = ""
        phase = "secret"
        finished = false
        started = false
        roles.clear()
        names.clear()
        seats.clear()
    }
}

File: harmony-host/shared/runtime/plugins/UndercoverPlugin.ts
import type { GamePlugin, Player, PluginResult } from "../types";
import { ok, err } from "../types";
import type { UndercoverWordStore } from "../UndercoverWordStore";

export class UndercoverPlugin implements GamePlugin {
  readonly id = "undercover";
  constructor(private store: UndercoverWordStore) {}

  private hostId: string | null = null;
  private undercoverCount = 1;
  private civilianWord = "";
  private undercoverWord = "";
  private phase: "secret" | "revealed" = "secret";
  private finished = false;
  private started = false;
  private roles = new Map<string, string>();
  private names = new Map<string, string>();
  private seats = new Map<string, number>();

  onStart(players: Player[], hostId: string): PluginResult {
    this.reset();
    const minNeed = this.store.minPlayers();
    if (players.length < minNeed) return err(`谁是卧底至少需要 ${minNeed} 人入座`);
    const count = this.store.undercoverCount();
    if (count < 1 || count >= players.length) return err(`卧底人数须在 1 到 ${players.length - 1} 之间，请先在配置里调整`);
    const pair = this.store.pickUnused();
    if (!pair) return err("词库没有可用词组，请先在配置里添加或重新加入");
    this.hostId = hostId; this.undercoverCount = count;
    this.civilianWord = pair.civilian; this.undercoverWord = pair.undercover;
    const shuffled = [...players].sort(() => Math.random() - 0.5);
    const ucIds = new Set(shuffled.slice(0, count).map((p) => p.id));
    for (const p of players) {
      this.roles.set(p.id, ucIds.has(p.id) ? "undercover" : "civilian");
      this.names.set(p.id, p.name); this.seats.set(p.id, p.seat ?? 0);
    }
    this.store.markUsed(pair.id);
    this.phase = "secret"; this.finished = false; this.started = true;
    return ok({ system: `词语已发放 · ${players.length} 人 · 卧底 ${count} 人 · 线下讨论后由房主揭晓` });
  }

  onAction(playerId: string, payload: any): PluginResult {
    if (!this.started) return err("对局未开始");
    if (this.finished || this.phase === "revealed") return err("已经揭晓");
    if (payload.op !== "reveal") return err("未知操作");
    if (playerId !== this.hostId) return err("只有房主可以揭晓");
    this.phase = "revealed"; this.finished = true;
    return ok({ finished: true, system: "身份已揭晓" });
  }

  snapshot(): any { return this.build(null); }
  snapshotFor(playerId: string): any { return this.build(playerId); }

  private build(viewerId: string | null): any {
    const revealed = this.phase === "revealed";
    const players: any[] = [];
    for (const [pid, role] of this.roles) {
      const o: any = { id: pid, name: this.names.get(pid) ?? "玩家", seat: this.seats.get(pid) ?? 0 };
      if (revealed) { o.role = role; o.word = role === "undercover" ? this.undercoverWord : this.civilianWord; }
      players.push(o);
    }
    const snap: any = { phase: this.phase, undercoverCount: this.undercoverCount, players, finished: this.finished };
    if (viewerId && this.roles.has(viewerId)) {
      const myRole = this.roles.get(viewerId)!;
      snap.myWord = myRole === "undercover" ? this.undercoverWord : this.civilianWord;
      if (revealed) snap.myRole = myRole;
    }
    if (revealed) { snap.civilianWord = this.civilianWord; snap.undercoverWord = this.undercoverWord; }
    return snap;
  }

  isFinished(): boolean { return this.finished; }
  reset(): void {
    this.hostId = null; this.undercoverCount = 1; this.civilianWord = ""; this.undercoverWord = "";
    this.phase = "secret"; this.finished = false; this.started = false;
    this.roles.clear(); this.names.clear(); this.seats.clear();
  }
}
